### PR TITLE
🎨 ⏱ Cleanup / optimise the server.init() function

### PR DESCRIPTION
**Pull request by ErisDS**
refs #2182

This is a set of commits that work towards making server/index.js a bit more readable.

Each commit has an explanation, most are minor changes, but 871598e and particularly 11171da are more interesting - moving code and behaviour to different locations.

There are many, many more changes that can be made here, but with this set, the flow of execution becomes very very easy to follow 😁

### CODE REVIEW

File: core/server/index.js [hunk 1]
```diff
@@ -77,6 +70,7 @@ function init() {
     }).then(function () {
         debug('Themes & apps done');
 
+        // Hmm... this doesn't return a promise
```
Comment (kirrg001) [inline core/server/index.js]: `models.init` operates synchronously.

File: core/server/data/db/health.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
+var KnexMigrator = require('knex-migrator'),
```
Comment (kirrg001) [inline core/server/data/db/health.js]: 👍

File: core/server/index.js [hunk 3]
```diff
@@ -63,17 +63,14 @@ function init() {
     // Initialize Internationalization
     i18n.init();
     debug('I18n done');
+    models.init();
```
Comment (kirrg001) [inline core/server/index.js]: ah okay 👍

File: core/server/models/settings.js [hunk 4]
```diff
@@ -163,17 +166,31 @@ Settings = ghostBookshelf.Model.extend({
...
+                isFirstRun = false;
 
             _.each(getDefaultSettings(), function each(defaultSetting, defaultSettingKey) {
                 var isMissingFromDB = usedKeys.indexOf(defaultSettingKey) === -1;
+
                 if (isMissingFromDB) {
-                    defaultSetting.value = defaultSetting.defaultValue;
+                    debug('isMissingFromDB', defaultSettingKey);
+                    // @TODO revisit this? maybe an even better way to do this with model events etc?
+                    if (defaultSettingKey === 'dbHash') {
```
Comment (kirrg001) [inline core/server/models/settings.js]: We could move this up to the `parseDefaultSettings` function, which makes it more readable i think.
We can define a default fallback object.

```
var fallback = {
   'dbHash': uuid.v4()
}
_.each(settings, function each(setting, settingName) {
  setting.value = setting.defaultValue || fallback(setting.key);
});
```

File: core/server/models/settings.js [hunk 5]
```diff
@@ -163,17 +166,31 @@ Settings = ghostBookshelf.Model.extend({
...
+                    }
+
                     insertOperations.push(Settings.forge(defaultSetting).save(null, options));
                 }
             });
 
-            return Promise.all(insertOperations);
+            return Promise.all(insertOperations).tap(function ifFirstRun() {
+                if (isFirstRun === true) {
+                    events.emit('server:first-run');
```
Comment (kirrg001) [inline core/server/models/settings.js]: Hmm not sure about this.
We would make the first server run dependent on initialising the settings. The event sounds very general.
I'd suggest removing it, if it's not used yet.

Comment (ErisDS) [inline core/server/models/settings.js]: Agree, removing

File: core/server/models/settings.js (hunk 4 above)
Comment (ErisDS) [inline core/server/models/settings.js]: Oh nice idea 👍

**Comment by ErisDS** [comment]
Have updated this with the requested changes

File: core/server/models/settings.js [hunk 6]
```diff
@@ -18,12 +17,18 @@ var Settings,
...
+        defaultSettingsFlattened = {},
+        dynamicDefault = {
+            dbHash: uuid.v4()
+        };
 
     _.each(defaultSettingsInCategories, function each(settings, categoryName) {
         _.each(settings, function each(setting, settingName) {
             setting.type = categoryName;
             setting.key = settingName;
+            if (dynamicDefault[setting.key]) {
```
Comment (ErisDS) [inline core/server/models/settings.js]: If I don't wrap this in an if, any empty settings end up with an undefined value, and triggering the `sqlite` does not support inserting default values.` error.

Comment (kirrg001) [inline core/server/models/settings.js]: 👍

**Review by kirrg001 — APPROVED** [review]

**Pull request merged**